common/src/main/java/dev/latvian/kubejs/server/FilterTagEventJS.java

```diff
@@ -0,0 +1,53 @@
+package dev.latvian.kubejs.server;
+
+import dev.latvian.kubejs.event.EventJS;
+import dev.latvian.kubejs.script.ScriptType;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.Tag;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class FilterTagEventJS<T> extends EventJS {
+	private final Map<ResourceLocation, Tag<T>> tagMap;
+	private final Function<ResourceLocation, Optional<T>> idToValue;
+
+	public FilterTagEventJS(Map<ResourceLocation, Tag<T>> tagMap, Function<ResourceLocation, Optional<T>> idToValue) {
+		this.tagMap = tagMap;
+		this.idToValue = idToValue;
+	}
+
+	private Tag<T> wrapTag(Tag<T> tag, Predicate<T> filter) {
+		return Tag.fromSet(tag.getValues().stream().filter(filter).collect(Collectors.toSet()));
```

Comment:
I feel like rather than wrapping the tag with each step and putting it back in the map, it'd be wiser to just join the predicates together and then apply them all at once for each tag *after* the event has been posted (see EventJS#afterPosted)

Reply:
I think it is easier to not do that instead, I wanted the predicates based on tags to be using the already filtered tags.